argoverse/utils/make_att_files.py

```diff
@@ -0,0 +1,365 @@
+
+
+def derivative(x: np.array) -> np.array:  # x: N-length np array
+    """
+    Compute derivative for velocity and acceleration 
+    """
+    x_tensor = torch.Tensor(x).unsqueeze(0).unsqueeze(0)
+    x_padded = torch.cat((x_tensor, (x_tensor[:, :, -1] - x_tensor[:, :, -2] + x_tensor[:, :, -1]).unsqueeze(0)), dim=2)
+    filters = torch.Tensor([-1, 1]).unsqueeze(0).unsqueeze(0)
+
+    return F.conv1d(x_padded, filters)[0, 0].numpy()
+
+
+def compute_v_a(traj: np.array) -> Tuple[np.array, np.array]:  # traj:Nx3
+    """
+    Compute velocity and acceleration
+    """
+
+    ind_valid = np.nonzero(1 - np.isnan(traj[:, 0]))[0]
+    dx = bspline_1d(ind_valid, traj[:, 0])
+    dy = bspline_1d(ind_valid, traj[:, 1])
+    dz = bspline_1d(ind_valid, traj[:, 2])
+
+    if len(ind_valid) > 0:
+        x_max = dx[ind_valid].max()
+        x_min = dx[ind_valid].min()
+        y_max = dy[ind_valid].max()
+        y_min = dy[ind_valid].min()
+
+        if np.sqrt((x_max - x_min) ** 2 + (y_max - y_min) ** 2) < 10:
+            return np.zeros_like(traj), np.zeros_like(traj)
+
+    vx = derivative(dx) * 10  # lidar freq 10 hz
+    vy = derivative(dy) * 10
+    vz = derivative(dz) * 10
+
+    ax = derivative(vx) * 10
+    ay = derivative(vy) * 10
+    az = derivative(vz) * 10
+
+    v = np.concatenate((vx[:, np.newaxis], vy[:, np.newaxis], vz[:, np.newaxis]), axis=1)
+    a = np.concatenate((ax[:, np.newaxis], ay[:, np.newaxis], az[:, np.newaxis]), axis=1)
+
+    v[np.isnan(traj[:, 0])] = np.nan
+    a[np.isnan(traj[:, 0])] = np.nan
+    return v, a
+
+
+if __name__ == "__main__":
+    # set root_dir to the correct path to your dataset folder
+    root_dir = "test_set/"
+    path_output_vis = "vis_output"
+    filename_output = "att_file.npy"
+
+    if not os.path.exists(path_output_vis):
+        os.mkdir(path_output_vis)
+
+    list_folders = ["test"]
+    list_name_class = ["VEHICLE", "PEDESTRIAN"]
+    count_track = 0
+    dict_att_all: Dict[str, Any] = {}
+    for name_folder in list_folders:
+
+        dict_att_all[name_folder] = {}
+        list_log_folders = glob.glob(os.path.join(root_dir, name_folder, "*"))
+        for ind_log, path_log in enumerate(list_log_folders):
+
+            id_log = path_log.split("/")[-1]
+            print("%s %s %d/%d" % (name_folder, id_log, ind_log, len(list_log_folders)))
+
+            if check_track_label_folder:
+                list_path_label_persweep = glob.glob(os.path.join(path_log, "per_sweep_annotations_amodal", "*"))
+                list_path_label_persweep.sort()
+
+                dict_track_labels: Dict[str, Any] = {}
+                for path_label_persweep in list_path_label_persweep:
+                    data = read_json_file(path_label_persweep)
+                    for data_obj in data:
+                        id_obj = data_obj["track_label_uuid"]
+
+                        if id_obj not in dict_track_labels.keys():
+                            dict_track_labels[id_obj] = []
```

Comment:
we can avoid the `if` branch in two ways :-) 
either use `collections.defaultdict`, or use `dict_track_labels.get(id_obj, [])`

Reply:
I think it's slightly different here. It's making an empty list for adding items in the future, not returning empty list when the key doesn't exist.

Reply:
oh nevermind .get() will only work for retrieving. For setting those keys 

```python
dict_track_labels.setdefault(id_obj, [])
```
should work

Reply:
"This method returns the key value available in the dictionary and if given key is not available then it will return provided default value."
https://www.tutorialspoint.com/python/dictionary_setdefault.htm